src/api/java/com/minecolonies/api/crafting/CompostRecipe.java

```diff
@@ -0,0 +1,182 @@
+
+    /**
+     * Get the input ingredient for this recipe (this is multiple alternative item types).
+     * Use {@link #getIngredients()} instead to also calculate the amount of items needed for one composting operation.
+     *
+     * @return The input.
+     */
+    public Ingredient getInput() { return this.input; }
+
+    /**
+     * Get the strength of this recipe.  A higher strength means fewer input items are required to craft the same
+     * quantity of output compost/dirt.
+     *
+     * @return The strength; typically 2, 4, or 8, but other values are possible.
+     */
+    public int getStrength() { return this.strength; }
+
+    /**
+     * The number of ticks that this recipe should take to ferment into compost.
+     * This is currently just for informational purposes and is not configurable (nor actually used).
+     *
+     * @return A number of ticks.
+     */
+    public int getFermentTime() { return FERMENT_TIME; }
+
+    @Override
+    public boolean matches(final IInventory inv, final World worldIn)
+    {
+        return false;
+    }
+
+    @NotNull
+    @Override
+    public ItemStack getCraftingResult(final IInventory inv)
+    {
+        return this.output.copy();
+    }
+
+    @Override
+    public boolean canFit(final int width, final int height)
+    {
+        return true;
+    }
+
+    @NotNull
+    @Override
+    public ItemStack getRecipeOutput()
+    {
+        return this.output;
+    }
+
+    @NotNull
+    @Override
+    public IRecipeSerializer<?> getSerializer()
+    {
+        return Serializer.INSTANCE;
+    }
+
+    /**
+     * Gets the list of ingredients for this recipe; there will only be one, but it is a multi-alternative stack
+     * and includes the specific count of items that are required.
+     *
+     * @return The list of input ingredients.
+     */
+    @NotNull
+    @Override
+    public NonNullList<Ingredient> getIngredients()
+    {
+        final NonNullList<Ingredient> ingredients = NonNullList.create();
+        ingredients.add(new CountedIngredient(this.input, calculateIngredientCount()));
+        return ingredients;
+    }
+
+    private int calculateIngredientCount()
+    {
+        return AbstractTileEntityBarrel.MAX_ITEMS / this.strength;
+    }
+
+    // JEI looks better if we render these as many individual recipes rather than as
+    // a few recipes with a massive number of alternative ingredients.
+    @NotNull
+    public static CompostRecipe individualize(@NotNull final Item item, @NotNull final CompostRecipe recipe)
+    {
+        return new CompostRecipe(recipe.getId(), Ingredient.fromItems(item), recipe.getStrength());
+    }
+
+    public static class Serializer extends ForgeRegistryEntry<IRecipeSerializer<?>>
+            implements IRecipeSerializer<CompostRecipe>
+    {
+        public static final Serializer INSTANCE = new Serializer();
+
+        static
+        {
+            INSTANCE.setRegistryName(CompostRecipe.ID);
+        }

```

Comment:
Please use a getter here instead of a direct field `getInstance()` is better here.

Reply:
This is the preferred style recommended by both Minecraft and Forge for recipe serializers.

Reply:
It might be the standard style forge and minecraft use. But not minecolonies. Static final fields are considered to be harder to unit test then static methods due to compiler inlineing.

Reply:
Another irrelevant argument.  There's no need to mock anything out, it can be tested in place without complication.  And the MineColonies source has no unit tests anyway.

Reply:
Translating this here: MineColonies tries to maintain a consistent style which might not be the default in Minecraft or Forge (similarly as our bracket style). So it makes sense to try to have the new code being added follow our default style as well.

Reply:
Style consistency is good (which is why I have tried to make all my code follow the brace style used elsewhere, despite it being unconventional for Java).  But rigorous adherence to a style checklist that contravenes good code design is not good.

There are some advantages to `getInstance()` -- the primary one is lazy instantiation, which is quite valuable in many cases, especially in a modloader context -- but not this case, since it's a stateless utility class, and it does impose a needless cost.  Unit testing is not a valid argument because either way it's static and can't be mocked out anyway (not that there is any need to) and because there aren't any.  Otherwise, a static final field is safer because it guarantees that the instance can't be tampered with (outside of extreme cases), and is automatically thread-safe (neither of which apply to getInstance).

There is technically an even better way to implement singletons in Java (via enums) but it's more obscure and the potential confusion didn't seem worth it since it wasn't used elsewhere in the codebase already.

Reply:
Through Enums is definitelty not worth it. Since the access seems solely within class (correct me if I'm wrong) I see no problem of having it potentially private and if we ever need to access it from the outside we create the getter.

Reply:
Almost, but not quite.  Registration of the serializer is the external usage, which happens over in CommonProxy.  It is technically possible to make that entirely private by having multiple internal subscribers to the mod event instead, but that's also a bit dodgy.

Reply:
For lazy instantiation you can use a static final Lazy<T> instance, which gives the best of both worlds.

Having said that, there is another legitimate use of getInstance -- which is how it is currently used throughout much of existing MineColonies -- that of allowing abstraction of the storage of the instance, for example by storing it inside another singleton instead of directly (e.g. how most stuff ends up indirected through the API instance).  That is a good design -- but it's also one that's not applicable here.